### ISSUE TITLE
Is it possible to vectorize samples in Monte Carlo estimation of ELBO gradients? 

### ISSUE THREAD

**Issue by lipingliulp** [blei-lab/edward#357]
In Monte Carlo estimation of ELBO gradients, samples are iterated in a for-loop. Is it possible to vectorize samples in the calculation?

Here is my thought based on my very basic understanding Edward. Suppose a single sample is tensor _**s**_ , and a array of samples is tensor _**S**_, which is one rank higher than _**s**_ and has the zero-th dimension corresponding to different samples. Then

1. many operations,  such as linear calculation and element-wise functions, on a single sample _**s**_  can be generalized to to _**S**_ ;
2. slicing of _**s**_ ( for Rao-Blackwellization) corresponds to slicing _**S**_ .

If some operations on _**s**_ cannot be generalized to _**S**_, can we postpone the "for" loop to the point before such operations? I am saying, we sample and try to calculate with _**S**_ at first; if we cannot calculate with _**S**_, we then iterate over its elements. 

I am not considering the problem that tensorflow does not support gradients of some operators, such as gather_n and scatter_update. I have no knowledge about how much speed-up we can get from the vectorization.

**Comment by altosaar** [comment]
It's not possible to get per-sample gradients in tensorflow right now. The tf.pack / tf.unpack solution is one way of doing it, as is creating a custom tf.get_variable that expands variables correctly. Here's a related thread: http://stackoverflow.com/questions/35731506/unaggregated-gradients-gradients-per-example-in-tensorflow

**Comment by dawenl** [comment]
Can you post some for-loop examples in Edward that you think can be vectorized? I think the general question to ask is whether this for-loop is executed in Python or something that got compiled in C (probably by tensorflow) -- the speedup for former will be substantial if done properly (but AFAIK will be hard to make generic), but if latter it will probably not worth the effect.

**Comment by lipingliulp** [comment]
Hi Dawen, thanks for joining this discussion. The following code is something in my mind. This snippet is revised from https://github.com/blei-lab/edward/blob/master/edward/inferences/klqp.py#L583 from line 583 to 621. I have marked the my assumptions in the code.  

I think the for loop in the original code is executed in python to construct the graph. I agree not all cases can be handled with vectors. Is it possible to run with vectors for the case we can handle? Please let me know more about your thought. 

```python

  # allocate space for log prob
  p_log_prob = tf.zeros([inference.n_samples])
  q_log_prob = tf.zeros([inference.n_samples])

  z_sample = sample_from_qz(inference.n_samples) # ASSUMPTION 1: a funtion sampling an array of samples for each z from qz
  for z, qz in six.iteritems(inference.latent_vars):
    # ASSUMPTION 2: z_sample[z] is the slice of samples for hidden variable z only
    # ASSUMPTION 3: qz.log_prob can calculate log prob for the entire array 
    z_log_prob = tf.reduce_sum(qz.log_prob(tf.stop_gradient(z_sample[z])), reduction_indices=[1:])
    if z in inference.scale:
      z_log_prob *= inference.scale[z]

    q_log_prob = z_log_prob

  if inference.model_wrapper is None:
    # Form dictionary in order to replace conditioning on prior or
    # observed variable with conditioning on a specific value.
    dict_swap = z_sample
    for x, qx in six.iteritems(inference.data):
      if isinstance(x, RandomVariable):
        if isinstance(qx, RandomVariable):
          qx_copy = copy(qx, scope=scope)
          dict_swap[x] = qx_copy.value()
        else:
          dict_swap[x] = qx

    for z in six.iterkeys(inference.latent_vars):
      z_copy = copy(z, dict_swap, scope=scope)
      # ASSUMPTION 4: z_copy.log_prob can calculate log prob for the entire array 
      z_log_prob = tf.reduce_sum(z_copy.log_prob(dict_swap[z]))
      if z in inference.scale:
        z_log_prob *= inference.scale[z]

      p_log_prob += z_log_prob

    for x in six.iteritems(inference.data):
      if isinstance(x, RandomVariable):
        x_copy = copy(x, dict_swap, scope=scope)
        # ASSUMPTION 5: the function can calculate log prob for each sample with the data
        x_log_prob = tf.reduce_sum(x_copy.log_prob(dict_swap[x]))
        if x in inference.scale:
          x_log_prob *= inference.scale[x]

        p_log_prob += x_log_prob

  else:
    x = inference.data
    # ASSUMPTION 5: the function can calculate log prob for each sample with the data
    # Inside this function, loop over z samples in this function if it is not feasible to calculate with tensor 
    p_log_prob = inference.model_wrapper.log_prob(x, z_sample)
```

**Comment by lipingliulp** [comment]
Hi Jaan,  sorry I do not understand your concern very well. The gradient calculation is still the same as 
https://github.com/blei-lab/edward/blob/master/edward/inferences/klqp.py#L638. 

Could you also check out the code in my answer to Dawen's comment? Thank you so much for putting thoughts in this issue.

**Comment by dawenl** [comment]
OK. I should have been more specific: I think it's fine to have for-loops executed in Python if it's not carrying out *actual* computation: constructing graph itself involves no actual computation. Executing the computational graph should be handled on the tensorflow level, which means for-loop should be OK (not sure about this though). @dustinvtran any better idea? 

@lipingliulp One thing that I recommend doing would be to vectorize the way you proposed for a specific case (so you don't need to consider how to make it generic) and benchmark with large `n_samples` and see if there is any substantial improvement.

**Comment by lipingliulp** [comment]
@dawenl Glad to know that this worth a try. 

I want to create benchmarks of tensorflow long time ago, and your suggestion is the last push: here is [one simple test](https://github.com/lipingliulp/benchmark-tensorflow/blob/master/forloop_vs_tensor.py) . It shows tensorflow takes significant time in handling for-loops. Please provide suggestions to make the test more like the calculation of Edward.

**Comment by dawenl** [comment]
Thanks for the benchmark! I varied `size` in your script and it looks like with list all the operation executions scale linearly with `size` and are much slower than the vectorized version, which is really bad -- this indicates if operations are in a list, it seems that tensorflow will have to execute them in a Python for-loop fashion. Furthermore, constructing graph with vectorization directly almost makes no time difference with different `size`. I think it is evident that the current Edward implementation will have scalability issues.

**Comment by dawenl** [comment]
To elaborate on the scalability issues:

From @lipingliulp's benchmark, here is what I learned: 

For intense computation, we have three options:

1. Vectorize it for both graph construction and (as a result) execution.

2. Have a for-loop like this: 
```
op_list = list()
for _ in range(n_ops):
    op_list.append(op) # some operation
```
But `n_ops` is not too large (say hundreds) and each `op` is vectorized, e.g. `tf.reduce_sum` etc.

3. Have the same for-loop as 2, but `n_ops` is large while each `op` is very simple, e.g. scalar addition. 

1 would be ideal, but I think it is gonna be very hard to make it generic. 3 is @lipingliulp's benchmark which is definitely not what we want. 2 seems to be a good compromise and is acceptable in most cases. I see Edward falls in 2. And it will depend a lot on the actual model/data. What do you guys think?

**Comment by lipingliulp** [comment]
@dawenl Maybe you can say more about what kind of op can be vectorized? Back to my original question, how do you think to vectorize samples?

**Comment by dustinvtran** [comment]
consider the following setup:
```python
x = Normal(mu=tf.zeros(5), sigma=tf.ones(5))
y = Normal(mu=x, sigma=tf.ones(5))
```
to generate a value of `y`, we call `y.value()`. the question is how we can generate multiple values of `y` without calling `y.value()` in a loop. in this case, it's immediate: simply draw a S x 5 matrix of values for `x`; this parameterizes a draw from `y` to also be a S x 5 matrix.

more generally, the problem is as follows: given that the user writes down a model, how do we vectorize all of its sampled operations? it's possible for specific models as the above. however, it doesn't work in non-trivial scenarios such as 
```python
x = tf.placeholder(tf.float32, [N, D])
beta = Normal(mu=tf.zeros(D), sigma=tf.ones(D))
y = Normal(mu=ed.dot(x, beta), sigma=tf.ones(N))
```
here, we cannot generate a matrix of samples `beta` and then somehow propagate it into the parameterization for `y` to get multiple samples at once. we'd have to somehow manipulate (and vectorize) the deterministic operations that a user defines in their model.

**Comment by dawenl** [comment]
@dustinvtran's example says all. 

I guess if S is not unreasonably large, it will be fine as long as the individual sample is vectorized, which I guess is the case here.

**Comment by lipingliulp** [comment]
Thanks for explanation. I think there is another opportunity to vectorize computation of factors of the same type. For example, in the mixture model, (x_i, z_i) and (x_j, z_j) are two different factors, but they use the same type of operation to compute. Many models are like this. I don't know how feasible it is. 

@dustinvtran, I agree with you in general. However, you example may not be valid: since normal distribution of tf support broadcasting of mu or sigma :)

**Comment by dustinvtran** [comment]
@lipingliulp: the dot product fails if we make `beta` a matrix (of multiple samples) rather than a vector (of one sample). re: your other point. we can do vectorization across data points already, no?

**Comment by dustinvtran** [comment]
that's right. it could be changed to `tf.matmul` but then it would no longer support matrix-vector multiplication
